File: frontend/src/types/attendance.ts
export type AttendanceStatus =
  | "PRESENT"
  | "ABSENT"
  | "HALF_DAY"
  | "LATE"
  | "ON_LEAVE"
  | "HOLIDAY";

export interface AttendanceRecord {
  id: string;
  employee_id: string;
  work_schedule_id?: string | null;
  attendance_date: string;
  check_in?: string | null;
  check_out?: string | null;
  expected_hours: number;
  worked_hours: number;
  overtime_hours: number;
  status: AttendanceStatus;
  notes?: string | null;
}

File: backend/app/schemas/attendance.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from app.models.attendance import AttendanceStatus


class AttendanceCreate(BaseModel):
    employee_id: str
    work_schedule_id: str | None = None
    attendance_date: date
    check_in: datetime | None = None
    check_out: datetime | None = None
    expected_hours: Decimal = Field(
        default=Decimal("0"),
        ge=0,
    )

    status: AttendanceStatus = AttendanceStatus.PRESENT
    notes: str | None = None

    @model_validator(mode="after")
    def validate_times(self):
        if (
            self.check_in is not None
            and self.check_out is not None
            and self.check_out < self.check_in
        ):
            raise ValueError("check_out cannot be before check_in")

        return self


class AttendanceUpdate(BaseModel):
    check_in: datetime | None = None
    check_out: datetime | None = None
    expected_hours: Decimal | None = Field(
        default=None,
        ge=0,
    )
    status: AttendanceStatus | None = None
    notes: str | None = None

    @model_validator(mode="after")
    def validate_times(self):
        if (
            self.check_in is not None
            and self.check_out is not None
            and self.check_out < self.check_in
        ):
            raise ValueError("check_out cannot be before check_in")

        return self


class AttendanceResponse(AttendanceCreate):
    id: str
    worked_hours: Decimal
    overtime_hours: Decimal
